File: python/src/sinetstream/config_client.py
#!/usr/bin/env python3

# Copyright (C) 2021 National Institute of Informatics
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import json
import logging
import os
import requests
import base64

from sinetstream.error import (
    NoConfigError, NoServiceError, InvalidConfigError, AuthorizationError, ConnectionError)

from sinetstream.utils import (
    user_config_dir,
    SecretValue)

import sinetstream.configs


logger = logging.getLogger(__name__)

insecure_log = os.getenv("SINETSTREAM_ENALBE_INSECURE_LOG") is not None


def test_enable_debug0():
    # https://gist.github.com/Daenyth/b57f8522b388e66fcf3b
    from http.client import HTTPConnection  # py3

    log = logging.getLogger('urllib3')
    log.setLevel(logging.DEBUG)

    # logging from urllib3 to console
    ch = logging.StreamHandler()
    ch.setLevel(logging.DEBUG)
    log.addHandler(ch)
    # print statements from `http.client.HTTPConnection` to console/stdout
    HTTPConnection.debuglevel = 1
# test_enable_debug0()


def test_enable_debug():
    # https://docs.python-requests.org/en/master/api/
    from http.client import HTTPConnection
    HTTPConnection.debuglevel = 1
    # logging.basicConfig() # you need to initialize logging,
    # otherwise you will not see anything from requests.
    # logging.getLogger().setLevel(logging.DEBUG)
    logger.setLevel(logging.DEBUG)
    requests_log = logging.getLogger("urllib3")
    requests_log.setLevel(logging.DEBUG)
    requests_log.propagate = True


if insecure_log:
    test_enable_debug()


user_auth_path = user_config_dir + "/auth.json"

api_path = "/api/v1"


def safe_gets(d, ks):
    v = d.get(ks[0])
    if len(ks) == 1:
        return v
    else:
        return safe_gets(v, ks[1:]) if isinstance(v, dict) else None


def get_value(d, k, t, optional=False, default=None):
    if k not in d:
        if optional:
            return default
        raise InvalidConfigError(f"No '{k}' in response")
    v = d.get(k)
    if not isinstance(v, t):
        if type(t) is not tuple:
            t = (t,)
        raise InvalidConfigError(f"'{k}' is not {','.join([x.__class__.__name__ for x in t])}")
    return v


def get_auth_info(auth_path=None):
    auth_path = auth_path if auth_path is not None else user_auth_path
    try:
        path = os.path.expanduser(auth_path)
        logger.debug(f"read {path}")
        with open(path) as fp:
            auth_info = json.load(fp)
            try:
                config_server = get_value(auth_info, "config-server", dict)
                address = get_value(config_server, "address", str)
                user = get_value(config_server, "user", str)
                secret_key = get_value(config_server, "secret-key", str)
            except InvalidConfigError as e:
                e0 = e.args[0].replace("in response", f"in {path}")  # hack...
                raise InvalidConfigError((e0,) + e.args[1:])
            return (address, user, secret_key)
    except FileNotFoundError:
        raise NoConfigError(f"No API key file exist: {auth_path}")


def get_access_token(session, base_uri, user, secret_key):
    resp = session.post(base_uri + "/authentication",
                        json={"user": user,
                              "secret-key": secret_key,
                              "strategy": "api-access"})
    logger.info(f"/authentication => {resp.status_code} {resp.text}")
    if resp.status_code == 201:
        pass
    else:
        logger.error(f"{resp.status_code} {resp.text}")
        if resp.status_code == 401:
            raise AuthorizationError(resp.text)
        else:
            raise ConnectionError(resp.text)
    res = resp.json()
    if insecure_log:
        logger.debug(f"resp.text={resp.text}")

    access_token = get_value(res, "accessToken", str)

    expired = safe_gets(res, ["authentication", "payload", "exp"])
    if insecure_log:
        import datetime
        if expired is not None:
            logger.debug(f"expired={expired}({datetime.datetime.fromtimestamp(expired)})")
        else:
            logger.debug("expired=None")

    return (access_token, expired)


def get_config_info(session, base_uri, common_headers, config_name):
    resp = session.get(base_uri + "/configs/" + config_name, headers=common_headers)
    logger.info(f"/configs/{config_name} => {resp.status_code} {resp.text}")
    if resp.status_code == 200:
        pass
    else:
        logger.error(f"{resp.status_code} {resp.text}")
        if resp.status_code == 401:
            raise AuthorizationError(resp.text)
        elif resp.status_code == 404:
            raise NoServiceError(resp.text)
        else:
            raise ConnectionError(resp.text)
    res = resp.json()
    if insecure_log:
        logger.error(f"{json.dumps(res, indent=2)}")
    # response is:
    # name: str
    # config:
    #     header:
    #         version: num
    #     config:
    #         service_name: dict
    # attachments:
    #     -
    #         target: str
    #         value: str
    # secrets:
    #     -
    #         target: str
    #         id: str

    if type(res) is not dict:
        raise InvalidConfigError("Bad response")
    name = get_value(res, "name", str)
    if name != config_name:
        raise InvalidConfigError(f"name '{name}' is not {config_name}")
    content = get_value(res, "config", dict)

    if "attachments" in res:
        attachments = get_value(res, "attachments", list)
        for attachment in attachments:
            if type(attachment) is not dict:
                raise InvalidConfigError("'attachments[]' is not dict")
    else:
        attachments = []

    if "secrets" in res:
        secrets = get_value(res, "secrets", list)
        for secret in secrets:
            if type(secret) is not dict:
                raise InvalidConfigError("'secrets[]' is not dict")
    else:
        secrets = []

    return (content, attachments, secrets)


def get_secret(session, base_uri, common_headers, sec_id, fingerprint):
    headers = common_headers.copy()
    headers["SINETStream-config-publickey"] = fingerprint
    resp = session.get(base_uri + "/secrets/" + sec_id, headers=headers)
    if resp.status_code == 200:
        pass
    else:
        logger.error(f"{resp.status_code} {resp.text}")
        if resp.status_code == 401:
            raise AuthorizationError(resp.text)
        elif resp.status_code == 404:
            raise NoServiceError(resp.text)
        else:
            raise ConnectionError(resp.text)
    res = resp.json()
    if insecure_log:
        logger.error(f"{json.dumps(res, indent=2)}")
    rid = get_value(res, "id", str)
    res_fingerprint = get_value(res, "fingerprint", str, optional=True)
    if res_fingerprint != fingerprint:
        logger.warning(f"fingerprint mismatch: {res_fingerprint}(server sent) != {fingerprint}(my pubkey)")
    target = get_value(res, "target", str)
    value64 = get_value(res, "value", str)
    if rid != sec_id:
        raise InvalidConfigError(f"id={rid} must be {sec_id}")
    value = base64.b64decode(value64)
    evalue = SecretValue(value, fingerprint)
    return (evalue, target)


def dict_update_value(d, path, value):
    for k in path[0:-1]:
        if k not in d:
            d[k] = {}
        if not isinstance(d[k], dict):
            raise InvalidConfigError(f"key {k} in the config file must be dict")
        d = d[k]
    d[path[-1]] = value


def get_config_params_server(service, config_name, mount_args=None, **kwargs):
    auth_info = get_auth_info(auth_path=kwargs.get("auth_path"))
    (address, user, secret_key) = auth_info
    if insecure_log:
        logger.debug(f"auth_info={(address, user, secret_key)}")

    base_uri = address + api_path

    session = requests.Session()
    if mount_args is not None:
        session.mount(*mount_args)

    (access_token, expired) = get_access_token(session, base_uri, user, secret_key)

    common_headers = {
        "Authorization": "Bearer " + access_token
    }

    (content, attachments, secrets) = get_config_info(session, base_uri, common_headers, config_name)

    header = get_value(content, "header", dict)
    config = get_value(content, "config", dict)

    version = get_value(header, "version", int)
    if version != 2:
        raise InvalidConfigError(f"version {version} must be 2")
    # fingerprint_header = get_value(header, "fingerprint", str, optional=True)

    if service is None:
        if len(config) == 0:
            raise NoConfigError("empty config")
        if len(config) > 1:
            logger.error(f"services in {config_name}: {[svc for svc in config.keys()]}")
            raise NoConfigError("The parameter service must be specified "
                                f"since many services are defined in the config {config_name}")
        service = list(config.keys()).pop()

    if service not in config:
        raise NoServiceError()

    import copy
    params = copy.deepcopy(get_value(config, service, dict))

    for attachment in attachments:
        target = get_value(attachment, "target", str)
        path = target.split(".")
        if len(path) < 2:
            raise InvalidConfigError(f"target {target} is too short")
        if path[0] == "*" or path[0] == service:
            value64 = get_value(attachment, "value", str)
            value = base64.b64decode(value64)
            dict_update_value(params, path[1:], value)

    for secret in secrets:
        target = secret.get("target")
        path = target.split(".")
        if path[0] == "*" or path[0] == service:
            ids = secret.get("ids")
            if ids is not None:
                # get the latest id
                sec_id = None
                max_ver = 0
                for e in ids:
                    id1 = e.get("id")
                    ver = e.get("version")
                    if id1 is None or ver is None:
                        raise InvalidConfigError(f"malformed entry '{e}' exists for '{target}'")
                    if ver > max_ver and id1 is not None:
                        max_ver = ver
                        sec_id = id1
                if sec_id is None:
                    raise InvalidConfigError(f"no valid id exists for '{target}'")
            else:
                sec_id = secret.get("id")
            _, _, fingerprint = sinetstream.configs.get_private_key()
            (evalue, sec_target) = get_secret(session, base_uri, common_headers, sec_id, fingerprint)
            if sec_target != target:
                raise InvalidConfigError(f"target={sec_target} must be {target}")
            dict_update_value(params, path[1:], evalue)

    # XXX: attachとsecretでtargetが被った場合はしらん。

    logger.debug(f"params={params}")
    return service, params

File: java/api/src/main/java/jp/ad/sinet/stream/utils/ConfigClient.java
/*
 * Copyright (C) 2021 National Institute of Informatics
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package jp.ad.sinet.stream.utils;

import jp.ad.sinet.stream.crypto.SecretDecoder;

import com.google.api.client.http.*;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.http.json.JsonHttpContent;
import com.google.api.client.json.GenericJson;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.JsonObjectParser;
import com.google.api.client.json.gson.GsonFactory;
import com.google.api.client.util.Key;
import com.google.gson.Gson;
import com.google.gson.annotations.SerializedName;
import jp.ad.sinet.stream.api.InvalidConfigurationException;
import jp.ad.sinet.stream.api.NoServiceException;
import lombok.extern.java.Log;

import java.io.IOException;
import java.io.Reader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


@Log
public class ConfigClient {
    public static class AuthConfigServer {
        public String address;
        public String user;
        @SerializedName("secret-key")
        public String secretKey;
        @SerializedName("expiration-date")
        public String expirationDate;
    }
    public static class AuthInfo {
        @SerializedName("config-server")
        public AuthConfigServer configServer;
    }
    static AuthInfo readAuthFile(Path authFile) {
        if (authFile == null)
            authFile = Paths.get(System.getProperty("user.home"), ".config", "sinetstream", "auth.json");
        try (Reader reader = Files.newBufferedReader(authFile)) {
            Gson gson = new Gson();
            return gson.fromJson(reader, AuthInfo.class);
        }
        catch (IOException e) {
            System.out.println(e);
            return null;
        }
    }
    private static class ConfigServer implements AutoCloseable {
        private HttpTransport httpTransport;
        private HttpRequestFactory httpRequestFactory;
        private JsonFactory jsonFactory;
        private GenericUrl baseUrl;
        private String accessToken;

        public ConfigServer(String address) {
            httpTransport = new NetHttpTransport();
            httpRequestFactory = httpTransport.createRequestFactory();
            jsonFactory = GsonFactory.getDefaultInstance();
            baseUrl = new GenericUrl(address + "/api/v1");
        }

        public ConfigServer(String address, HttpTransport httpTransport) {
            this.httpTransport = httpTransport;
            httpRequestFactory = httpTransport.createRequestFactory();
            jsonFactory = GsonFactory.getDefaultInstance();
            baseUrl = new GenericUrl(address + "/api/v1");
        }

        // https://qiita.com/alt_yamamoto/items/0d72276c80589493ceb4
        // https://github.com/googleapis/google-http-java-client
        public static class AuthenticationResUser extends GenericJson {
            @Key
            public long id;
            @Key
            public String name;
        }
        public static class AuthenticationRes extends GenericJson {
            @Key
            public String accessToken;
            @Key
            public AuthenticationResUser user;
        }
        public void postAuthentication(String user, String accessKey) throws IOException {
            GenericUrl url = baseUrl.clone();
            url.appendRawPath("/authentication");

            Map<String, Object> params = new java.util.HashMap<>();
            params.put("strategy", "api-access");
            params.put("user", user);
            params.put("secret-key", accessKey);

            try {
                HttpRequest hreq = httpRequestFactory.buildPostRequest(url, new JsonHttpContent(jsonFactory, params));
                hreq.setParser(new JsonObjectParser(GsonFactory.getDefaultInstance()));
                HttpResponse hres = hreq.execute();
                try {
                    log.fine(hres.getStatusCode() + " " + hres.getStatusMessage());
                    log.fine(hres.getContentType());
                    if (hres.getStatusCode() != 201) {
                        throw new InvalidConfigurationException("server failure: " + hres.getStatusCode());
                    }
                    AuthenticationRes res = hres.parseAs(AuthenticationRes.class);
                    accessToken = res.accessToken;
                }
                finally {
                    hres.disconnect();
                }
            }
            catch (IOException e){
                System.out.println(e);
                throw e;
            }
        }
        public void close() {
            try {
                httpTransport.shutdown();
            }
            catch (IOException e) {
                System.out.println(e);
            }
        }
        public static class ConfigsRes extends GenericJson {
            public static class Config extends GenericJson {
                public static class Header extends GenericJson {
                    @Key
                    public long version;
                    @Key
                    public String fingerprint;
                }
                public static class ConfigConfig extends GenericJson {
                }

                @Key
                public Header header;
                @Key
                public ConfigConfig config;
            }
            public static class Attachment extends GenericJson {
                @Key
                public String value;
                @Key
                public String target;
            }
            public static class Secret extends GenericJson {
                public static class SId extends GenericJson {
                    @Key
                    public String id;
                    @Key
                    public long version;
                }
                @Key
                public List<SId> ids;
                @Key
                public String id;
                @Key
                public String target;
            }

            @Key
            public String name;
            @Key
            public Config config;
            @Key
            public List<Attachment> attachments;
            @Key
            public List<Secret> secrets;
        }
        public ConfigsRes getConfigs(String configName) {
            GenericUrl url = baseUrl.clone();
            url.appendRawPath("/configs/" + configName);

            try {
                HttpRequest hreq = httpRequestFactory.buildGetRequest(url);
                HttpHeaders headers = new HttpHeaders();
                headers.setAuthorization("Bearer " + accessToken);
                hreq.setHeaders(headers);
                hreq.setParser(new JsonObjectParser(GsonFactory.getDefaultInstance()));
                HttpResponse hres = hreq.execute();
                try {
                    log.fine(hres.getStatusCode() + " " + hres.getStatusMessage());
                    log.fine(hres.getContentType());
                    if (hres.getStatusCode() != 200) {
                        throw new InvalidConfigurationException("server failure: " + hres.getStatusCode());
                    }
                    ConfigsRes res = hres.parseAs(ConfigsRes.class);
                    return res;
                }
                finally {
                    hres.disconnect();
                }
            }
            catch (IOException e){
                System.out.println(e);
                throw new InvalidConfigurationException(e);
            }
        }

        public static class SecretsRes extends GenericJson {
            @Key
            public String id;
            @Key
            public String fingerprint;
            @Key
            public String target;
            @Key
            public String value;
        }
        public SecretsRes getSecret(String secId, String fingerprint) {
            GenericUrl url = baseUrl.clone();
            url.appendRawPath("/secrets/" + secId);

            try {
                HttpRequest hreq = httpRequestFactory.buildGetRequest(url);
                HttpHeaders headers = new HttpHeaders();
                headers.setAuthorization("Bearer " + accessToken);
                if (fingerprint != null) {
                    headers.set("SINETStream-config-publickey", fingerprint);
                }
                hreq.setHeaders(headers);
                hreq.setParser(new JsonObjectParser(GsonFactory.getDefaultInstance()));
                HttpResponse hres = hreq.execute();
                try {
                    log.fine(hres.getStatusCode() + " " + hres.getStatusMessage());
                    log.fine(hres.getContentType());
                    if (hres.getStatusCode() != 200) {
                        throw new InvalidConfigurationException("server failure: " + hres.getStatusCode());
                    }
                    SecretsRes res = hres.parseAs(SecretsRes.class);
                    return res;
                }
                finally {
                    hres.disconnect();
                }
            }
            catch (IOException e){
                System.out.println(e);
                throw new InvalidConfigurationException(e);
            }
        }
    }
    @SuppressWarnings("unchecked")
    private static <T> void update_value(Map<String,Object> params, String[] tpath, T value) {
        for (int i = 1; i < tpath.length - 1; i++) {
            String k = tpath[i];
            if (!params.containsKey(k))
                params.put(k, new HashMap<String,Object>());
            Object child = params.get(k);
            if (!(child instanceof Map)) {
                throw new InvalidConfigurationException("key " + k + " in the config must be Map<>");
            }
            params = (Map<String,Object>) child;
        }
        params.put(tpath[tpath.length - 1], value);
    }
    @SuppressWarnings("unchecked")
    public static Map<String, Object> getConfig(String serviceName, String configName, boolean needAllKey, Path authFile, Path privKeyFile, Object debugHttpTransport) {
        AuthInfo authInfo = readAuthFile(authFile);
        HttpTransport httpTransport = debugHttpTransport != null ? (HttpTransport) debugHttpTransport : new NetHttpTransport();
        return getConfig(serviceName, configName, needAllKey, authInfo, privKeyFile, httpTransport);
    }
    @SuppressWarnings("unchecked")
    public static Map<String, Object> getConfig(String serviceName, String configName, boolean needAllKey, AuthInfo authInfo, Path privKeyFile, HttpTransport httpTransport) {
        try (ConfigServer configServer = new ConfigServer(authInfo.configServer.address, httpTransport)) {
            configServer.postAuthentication(authInfo.configServer.user, authInfo.configServer.secretKey);
            ConfigServer.ConfigsRes configs = configServer.getConfigs(configName);
            log.fine("XXX:configs=" + configs);
            if (configs.name == null) {
                throw new InvalidConfigurationException("server must return name");
            }
            if (!configs.name.equals(configName)) {
                throw new InvalidConfigurationException("server must return name=" + configName + ", but " + configs.name);
            }
            Map<String,Object> params;
            if (configs.config.header != null) {
                // if the header section exists, assume version >= 2.
                if (configs.config.header.version != 2) {
                    throw new InvalidConfigurationException("config verison must be 2, but " + configs.config.header.version);
                }
                if (configs.config.config == null) {
                    throw new InvalidConfigurationException("config section does not exist");
                }
                params = (Map<String,Object>) MessageUtils.getServiceParam((Map<String,Object>) configs.config.config, serviceName);
            } else {
                // assume version 1
                params = (Map<String,Object>) MessageUtils.getServiceParam(configs.config, serviceName);
            }
            if (params == null) {
                throw new NoServiceException();
            }

            // expand attachments
            if (configs.attachments != null) {
                Base64.Decoder b64decoder = Base64.getDecoder();
                for (ConfigServer.ConfigsRes.Attachment attachment : configs.attachments) {
                    String[] tpath = attachment.target.split("[.]");
                    if (tpath.length < 2) {
                        throw new InvalidConfigurationException("target " +  attachment.target + " is too short");
                    }
                    if (tpath[0].equals("*") || tpath[0].equals(serviceName)) {
                        byte[] value = b64decoder.decode(attachment.value);
                        update_value(params, tpath, value);
                    }
                }
            }

            // expand secrets
            if (configs.secrets != null) {
                Base64.Decoder b64decoder = Base64.getDecoder();
                SecretDecoder secretDecoder = null;
                String fp = null;
                for (ConfigServer.ConfigsRes.Secret secret : configs.secrets) {
                    String[] tpath = secret.target.split("[.]");
                    if (tpath.length < 2) {
                        throw new InvalidConfigurationException("target " +  secret.target + " is too short");
                    }
                    if (tpath[0].equals("*") || tpath[0].equals(serviceName)) {
                        if (secretDecoder == null) {
                            try {
                                secretDecoder = new SecretDecoder(privKeyFile);
                                secretDecoder.setupPrivKey();
                                fp = secretDecoder.getFingerprint();
                            }
                            catch (Exception e) {
                                log.warning("config failure: " + e);
                                throw new RuntimeException(e);
                            }
                        }
                        if (pathCompare(tpath, 1, "crypto", "key")) {
                            Map<Integer, SecretValue> keys = new HashMap<>();
                            // writer: read max version on "crypto.key", set to "crypto._keys"
                            // reader: read all version on "crypto.key", set to "crypto._keys"
                            long maxVer = 0;
                            for (ConfigServer.ConfigsRes.Secret.SId sid : secret.ids) {
                                if (!needAllKey && sid.version <= maxVer)
                                    continue;
                                ConfigServer.SecretsRes sec = configServer.getSecret(sid.id, fp);
                                if (!sec.id.equals(sid.id) || (fp != null && !sec.fingerprint.equals("SHA256:" + fp)) || !sec.target.equals(secret.target)) {
                                    throw new InvalidConfigurationException("server returns malformed response: " + sec);
                                }
                                if (!needAllKey)
                                    keys.clear();
                                keys.put((int)sid.version, new SecretValue(b64decoder.decode(sec.value), sec.fingerprint, secretDecoder));
                            }
                            if (keys.isEmpty())
                                throw new InvalidConfigurationException("no valid id exists for " + secret.target);
                            String[] crypto_keys = { "<dummy>", "crypto", "_keys" };
                            update_value(params, crypto_keys, keys);
                        } else {
                            String secId = null;
                            if (secret.ids != null) {
                                long maxVer = 0;
                                for (ConfigServer.ConfigsRes.Secret.SId sid : secret.ids) {
                                    if (sid.version > maxVer && sid.id != null) {
                                        secId = sid.id;
                                        maxVer = sid.version;
                                    }
                                }
                            } else if (secret.id != null) {
                                secId = secret.id;
                            }
                            if (secId == null) {
                                throw new InvalidConfigurationException("no valid id exists for " + secret.target);
                            }
                            ConfigServer.SecretsRes sec = configServer.getSecret(secId, fp);
                            if (!sec.id.equals(secId) || (fp != null && !sec.fingerprint.equals("SHA256:" + fp)) || !sec.target.equals(secret.target)) {
                                throw new InvalidConfigurationException("server returns malformed response: " + sec);
                            }
                            update_value(params, tpath, new SecretValue(b64decoder.decode(sec.value), sec.fingerprint, secretDecoder));
                        }
                    }
                }
            }
            return  params;
        }
        catch (RuntimeException | IOException e) {
            log.warning("config failure: " + e);
            //e.printStackTrace();
            return null;
        }
    }

    static boolean pathCompare(String[] tpath, int index, String... ss) {
        int end = tpath.length;
        if (end <= index)
            return false;
        if (end - index != ss.length)
            return false;
        int k = 0;
        for (int i = index; i < end; i++, k++) {
            if (!tpath[i].equals(ss[k]))
                return false;
            }
        return true;
    }
}
